### PR TITLE
- tools.PkgConfig helper

### PR DESCRIPTION
**Pull request by SSE4**
tool.PkgConfig helper to obtain various properties from pkg-config

### CODE REVIEW

**Comment by lasote** [comment]
Please, could you explain a little the use case behind this PR? Thanks!

**Comment by SSE4** [comment]
@lasote 
certain libraries (e.g. SRT) accepts variables like OPENSSL_ROOT_DIR/OPENSSL_LIBRARIES/OPENSSL_INCLUDE_DIR which specify options how to link with dependencies (e.g. OpenSSL in case of SRT).
given pkg-config and pc files, you can set them like:
```
with tools.environment_append({'PKG_CONFIG_PATH': 'C:\\SRT\\lib\\pc'}):
  pc = tools.PkgConfig('openssl')
  cmake.definitions['OPENSSL_ROOT_DIR'] = pc.prefix
  cmake.definitions['OPENSSL_LIBRARIES'] = pc.libs
```

**Comment by lasote** [comment]
So the main use case is to link against system libraries, right? You can get that information from conan packages that you are depending on.

**Comment by SSE4** [comment]
yep, it's for linking against system libraries which provide pkg-config (pc files)

File: conans/client/tools/pkg_config.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import subprocess
+
+_global_output = None
+
+
+def cmd_output(command):
```
Comment (lasote) [inline conans/client/tools/pkg_config.py]: Same here.

File: conans/client/tools/pkg_config.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        variable_names = self._parse_output('print-variables').split()
+        self._variables = dict()
+        for name in variable_names:
+            self.variables[name] = self._parse_output('variable=%s' % name)
+
+        self.info = dict()
+        for option in ['cflags', 'cflags-only-I', 'cflags-only-other',
+                       'libs', 'libs-only-L', 'libs-only-l', 'libs-only-other',
+                       'print-provides', 'print-requires', 'print-requires-private']:
+            self.info[option] = self._parse_output(option).split()
```
Comment (lasote) [inline conans/client/tools/pkg_config.py]: Probably is a better idea to not invoke all the possible options and just to it lazy in the @property methods.
Something like:
```
@property
def cflags(self):
    if not "cflags" in self.info:
         # Call the command and assign to self.info
    return self.info['cflags']
```

Comment (SSE4) [reply conans/client/tools/pkg_config.py]: yep, good suggestion

File: conans/client/tools/pkg_config.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import subprocess
+
+_global_output = None
```
Comment (lasote) [inline conans/client/tools/pkg_config.py]: Not used.

**Comment by SSE4** [comment]
@lasote I've made implementation lazy and corrected minor issues

**Comment by SSE4** [comment]
another use-case I've just realised today: there are libraries that produce .pc after build (such as ffmpeg). it's much better to populate **deps_cpp_info** with information from pkg-config, rather maintaining it manually in the conanfile.

**Comment by lasote** [comment]
The problem with that use case is that the PREFIX has to be adjusted because otherwise a fixed path from the machine where the package was generated will be populated in every consumer.
Maybe the helper could provide some way to override the prefix with the current package_folder?

If possible, the docs [here](http://docs.conan.io/en/latest/integrations/pkg_config_pc_files.html) should be completed with the new approaches, probably prioritizing if it's cleaner than the already proposed.

**Comment by SSE4** [comment]
as it is mentioned in docs, it's possible to use/pass `–define-prefix` into pkg-config
I am going to add this to PR
anything else need to be done in order to get PR applied?

**Comment by SSE4** [comment]
@lasote I've added an additional test case which demonstrates how to override prefix.
actually, better to use `--define-variable=prefix=/home/conan` rather `--define-prefix` (which uses location of pc file), for instance:
`pkg_config = PkgConfig("libastral", variables={'prefix': '/home/conan'})`

**Comment by lasote** [comment]
Ok, will take a look for the 0.29. 
0.28 is packed and ready to release.

**Comment by memsharded** [comment]
Good contribution, merging it, thanks! Will be in 0.29

**Pull request merged**